5.0 Draft/IDD/IDDs Service Registry/eu.arrowhead.service-discovery-http-json.yml

```diff
@@ -0,0 +1,631 @@
+  x-semantics: to-add
+  x-Ontology: to-add
+
+paths:
+
+  /echo:
+    get:  
+      summary: Endpoint to check the core system availability. 
+      description: Returns an OK to indicate that the core system is available. 
+      responses:
+        '200':
+          description: OK
+
+  /services:
+
+    get:
+      operationId: query
+      description: |
+        Searches for records matching any query parameters. If no query parameters are specified,
+        any records the consuming system are permitted to access may be provided in the response.
+
+        If the `limit` query parameter is used and given a value larger than 16, or if the `offset`
+        query parameter is used with any value, the operation is escalated from type `query` to type
+        `scan`. Permission to use the latter operation is intended only for consumers that need to
+        interact with larger number of service providers, that analyze larger numbers of records,
+        that list them for administrators or operators, or that have other special needs.
+
+        As long as the operation remains a plain `query`, it will always limit its response to 16
+        records, unless a smaller `limit` is specified. If escalated to type `scan`, default and
+        maximum limits depend on how the service is configured. The limits may vary depending on the
+        attributes of the consumer, such as by allowing for a higher limit if the consumer is an
+        administrator.
+
+        Any matching records a consumer are not allowed to access will be silently discarded before
+        the record limit is imposed.
+      parameters:
+        - $ref: '#/components/parameters/query-limit'
+        - $ref: '#/components/parameters/query-offset'
+        - $ref: '#/components/parameters/query-q'
+      responses:
+        '200':
+          description: Matching service records.
+          content:
+            application/json:
+              schema:
+                type: array
+                items: { $ref: '#/components/schemas/Record' }
+          headers:
+            Cache-Control:
+              $ref: '#/components/headers/Cache-Control'
+            Limit:
+              $ref: '#/components/headers/Limit'
+        '400': { $ref: '#/components/responses/400-BadRequest-ExpressionIsInvalid' }
+        '401': { $ref: '#/components/responses/401-Unauthorized' }
+        '403': { $ref: '#/components/responses/403-Forbidden' }
+        '408': { $ref: '#/components/responses/408-RequestTimeout' }
+        '414': { $ref: '#/components/responses/414-URITooLarge' }
+        '415': { $ref: '#/components/responses/415-UnsupportedMediaType' }
+        '429': { $ref: '#/components/responses/429-TooManyRequests' }
+        '431': { $ref: '#/components/responses/431-RequestHeaderFieldsTooLarge' }
+        '500': { $ref: '#/components/responses/500-InternalServerError' }
+        '503': { $ref: '#/components/responses/503-ServiceUnavailable' }
+
+    patch:
```

Comment:
Why patch? A POST with idempotency would be more "REST" imho. A PUT as an "upsert" would work as well but then the ID needs to be in the path.

Reply:
This is a batch operation, which means that the entire collection of records is being updated. A PUT would mean that all of them are replaced, and a POST would mean that we are adding an array to a list of arrays, which is not true. I figured PATCH would be a good way of signaling that something else is going on here, which is that some records are updated, some are added and some are left as they are.

Reply:
Now that I see, should the ability to register/modify services be here to begin with? As I see it that is task of the service registry administration interface, rather than the service discovery. And as of now, it is present in both interfaces. Let me know what you think, but maybe having just the administrative interface for this kind of registration and modifications is enough.

Reply:
I think we should limit what is possible to do in the service-discovery interface. Registering and unregistering a service for me should be the only things a provider should be able to do using this interface. The modifying existing entries I think should be fully part of the service registry administration interface. This means that this operation can remain a idempotent POST. While in the service registry administration interface we still have the Add, update, and delete operations, so no need to try to fit update and create in the same operation. What do you think about it?